In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, classification_report
from sklearn.preprocessing import StandardScaler

In [2]:
df = pd.read_csv("novagen_dataset.csv")

In [3]:
df.head()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False


In [4]:
bool_col = ["Diet_Type__Vegan","Diet_Type__Vegetarian", "Blood_Group_AB", "Blood_Group_B", "Blood_Group_O"]
df[bool_col] = df[bool_col].astype(int)

In [5]:
# Split features and target
X = df.drop("Target", axis=1)
y = df["Target"]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
# Scaling
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [29]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

lrm = LogisticRegression(max_iter=1000)
lrm.fit(X_train_scaled, y_train)

y_pred_train = lrm.predict(X_train_scaled)
y_pred_test = lrm.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.8151590522319675
Test Accuracy:  0.8225130890052356
Train Recall :  0.8359788359788359
Test Recall :  0.8376237623762376
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [85]:
# KNN Classifier
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=7, metric="euclidean")
knn.fit(X_train_scaled, y_train)

y_pred_train = knn.predict(X_train_scaled)
y_pred_test = knn.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9219793166644848
Test Accuracy:  0.8806282722513089
Train Recall :  0.926681783824641
Test Recall :  0.8831683168316832
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [37]:
# Naive Bayes
from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()
nb.fit(X_train_scaled, y_train)

y_pred_train = nb.predict(X_train_scaled)
y_pred_test = nb.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.81149365100144
Test Accuracy:  0.8209424083769633
Train Recall :  0.8062484252960443
Test Recall :  0.80990099009901
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [88]:
# Decision Tree
from sklearn.tree import DecisionTreeClassifier

dtc = DecisionTreeClassifier(max_depth=7)
dtc.fit(X_train_scaled, y_train)

y_pred_train = dtc.predict(X_train_scaled)
y_pred_test = dtc.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9007723524021469
Test Accuracy:  0.8900523560209425
Train Recall :  0.9536407155454775
Test Recall :  0.9336633663366337
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [89]:
# SVC
from sklearn.svm import SVC

svc = SVC()
svc.fit(X_train_scaled, y_train)

y_pred_train = svc.predict(X_train_scaled)
y_pred_test = svc.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9509098049482917
Test Accuracy:  0.9335078534031414
Train Recall :  0.9627110103300579
Test Recall :  0.9425742574257425
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [91]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(n_estimators=200, max_depth=9, random_state=42)

rf.fit(X_train_scaled, y_train)

y_pred_train = rf.predict(X_train_scaled)
y_pred_test = rf.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9498625474538552
Test Accuracy:  0.9287958115183246
Train Recall :  0.9730410682791635
Test Recall :  0.9475247524752475
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [97]:
# Ensemble Gradient Boosting Classifier
from sklearn.ensemble import GradientBoostingClassifier, VotingClassifier

gb = GradientBoostingClassifier(n_estimators=200, learning_rate=0.1, max_depth=4, random_state=42)

gb.fit(X_train_scaled, y_train)

y_pred_train = gb.predict(X_train_scaled)
y_pred_test = gb.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9664877601780337
Test Accuracy:  0.9418848167539267
Train Recall :  0.9727891156462585
Test Recall :  0.9455445544554455
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



In [98]:
# Ensemble Voting Classifier
from sklearn.ensemble import VotingClassifier


voting_clf = VotingClassifier(
    estimators=[
        ("lr", LogisticRegression(max_iter=1000)),
        ("knn", KNeighborsClassifier(n_neighbors=7)),
        ("rf", RandomForestClassifier(n_estimators=200, random_state=42))
    ],
)

voting_clf.fit(X_train_scaled, y_train)

y_pred_train = voting_clf.predict(X_train_scaled)
y_pred_test = voting_clf.predict(X_test_scaled)

print("Train Accuracy: ",accuracy_score(y_train, y_pred_train))
print("Test Accuracy: ",accuracy_score(y_test, y_pred_test))
print("Train Recall : ",recall_score(y_train, y_pred_train))
print("Test Recall : ",recall_score(y_test, y_pred_test))
print("Classification Report\n",classification_report(y_test, y_pred))

Train Accuracy:  0.9569315355413012
Test Accuracy:  0.9157068062827225
Train Recall :  0.9674981103552532
Test Recall :  0.9168316831683169
Classification Report
               precision    recall  f1-score   support

           0       0.88      0.91      0.90       900
           1       0.92      0.89      0.91      1010

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 83.7%  |
| KNN                  | 88.3%  |
| Random Forest        | 94.7%  |
| Gradient Boosting    | 94.5%  |
| Voting Classifier    | 93.07% |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 93.7%